File: joern_server/graph/scala_parse.py
def _extract_scala_tuples(stdout: str) -> list[str]:
    """Extract individual tuple strings from Scala `List(...)` output."""
    if not stdout:
        return []
    # Find "= List(" — the value list, skipping the type-annotation List(...)
    idx = stdout.find("= List(")
    if idx != -1:
        content_start = idx + 7
    else:
        idx = stdout.find("List(")
        if idx == -1:
            return []
        content_start = idx + 5
    depth = 1
    pos = content_start
    while pos < len(stdout) and depth > 0:
        if stdout[pos] == "(":
            depth += 1
        elif stdout[pos] == ")":
            depth -= 1
        pos += 1
    content = stdout[content_start:pos - 1]
    tuples: list[str] = []
    i = 0
    while i < len(content):
        if content[i] == "(":
            d = 1
            j = i + 1
            while j < len(content) and d > 0:
                if content[j] == "(":
                    d += 1
                elif content[j] == ")":
                    d -= 1
                j += 1
            if d == 0:
                tuples.append(content[i:j])
                i = j
                continue
        i += 1
    return tuples


def _split_scala_tuple(tuple_str: str) -> list[str]:
    """Split a Scala tuple string by top-level commas into fields."""
    inner = tuple_str[1:-1] if tuple_str.startswith("(") and tuple_str.endswith(")") else tuple_str
    fields: list[str] = []
    current: list[str] = []
    depth = 0
    in_string = False
    escaped = False
    i = 0
    while i < len(inner):
        ch = inner[i]
        if escaped:
            current.append(ch)
            escaped = False
            i += 1
            continue
        if ch == "\\":
            current.append(ch)
            escaped = True
            i += 1
            continue
        if ch == '"':
            in_string = not in_string
            current.append(ch)
            i += 1
            continue
        if in_string:
            current.append(ch)
            i += 1
            continue
        if ch in "([{":
            depth += 1
            current.append(ch)
            i += 1
            continue
        if ch in ")]}":
            depth -= 1
            current.append(ch)
            i += 1
            continue
        if ch == "," and depth == 0:
            fields.append("".join(current).strip())
            current = []
            i += 1
            continue
        current.append(ch)
        i += 1
    if current:
        fields.append("".join(current).strip())
    return fields


def _parse_scala_field(field: str):
    """Parse a single Scala value (string, int, Some(value=...), None) into Python."""
    field = field.strip()
    if not field:
        return None
    # Strip trailing "L" suffix from Scala Long literals
    long_suffix = False
    if field.endswith("L") and len(field) > 1:
        rest = field[:-1]
        if rest.isdigit() or (rest.startswith("-") and rest[1:].isdigit()):
            field = rest
            long_suffix = True
    if field.startswith('"') and field.endswith('"') and len(field) >= 2:
        inner = field[1:-1]
        return inner.replace('\\"', '"').replace("\\\\", "\\")
    if field == "None":
        return None
    if field.startswith("Some(") and field.endswith(")"):
        inner_val = field[5:-1].strip()
        # Handle `Some(value = 42)` format from Option[Int] output
        eq_idx = inner_val.find(" = ")
        if eq_idx != -1:
            num_str = inner_val[eq_idx + 3:].strip()
            try:
                return int(num_str)
            except ValueError:
                return inner_val
        try:
            return int(inner_val)
        except ValueError:
            return inner_val
    try:
        return int(field)
    except ValueError:
        return field


def _parse_metadata_tuples(stdout: str) -> dict[str, dict]:
    """Parse 6-field metadata tuples (id, code, line, column, order, label) into {nodeId: metadata}."""
    metadata: dict[str, dict] = {}
    for t in _extract_scala_tuples(stdout):
        fields = _split_scala_tuple(t)
        if len(fields) < 6:
            continue
        try:
            fid = _parse_scala_field(fields[0])
            code = _parse_scala_field(fields[1])
            line_num = _parse_scala_field(fields[2])
            col_num = _parse_scala_field(fields[3])
            order = _parse_scala_field(fields[4])
            node_type = _parse_scala_field(fields[5])
            metadata[str(fid)] = {
                "code": code if isinstance(code, str) else str(code) if code is not None else "",
                "line_number": line_num,
                "column_number": col_num,
                "order": order if order is not None else -1,
                "argument_index": -1,
                "node_type": node_type if isinstance(node_type, str) else "",
            }
        except Exception:
            continue
    return metadata


def _parse_ast_tuples(stdout: str) -> tuple[list[dict], list[dict], dict[str, dict]]:
    """Parse 7-field AST tuples (id, code, line, column, order, label, parentId) into (nodes, edges, metadata)."""
    nodes: list[dict] = []
    edges: list[dict] = []
    metadata: dict[str, dict] = {}
    for t in _extract_scala_tuples(stdout):
        fields = _split_scala_tuple(t)
        if len(fields) < 7:
            continue
        try:
            fid = _parse_scala_field(fields[0])
            code = _parse_scala_field(fields[1])
            line_num = _parse_scala_field(fields[2])
            col_num = _parse_scala_field(fields[3])
            order = _parse_scala_field(fields[4])
            node_type = _parse_scala_field(fields[5])
            parent_id = _parse_scala_field(fields[6])
            node_id_str = str(fid)
            nodes.append({
                "id": node_id_str,
                "label": node_type if isinstance(node_type, str) else str(node_type) if node_type is not None else "",
            })
            metadata[node_id_str] = {
                "code": code if isinstance(code, str) else str(code) if code is not None else "",
                "line_number": line_num,
                "column_number": col_num,
                "order": order if order is not None else -1,
                "argument_index": -1,
                "node_type": node_type if isinstance(node_type, str) else "",
            }
            if parent_id is not None:
                edges.append({
                    "source": str(parent_id),
                    "target": node_id_str,
                    "label": "",
                })
        except Exception:
            continue

    node_ids = {n["id"] for n in nodes}
    for e in edges:
        for nid in (e["source"], e["target"]):
            if nid not in node_ids:
                nodes.append({"id": nid, "label": "", "shape": ""})
                metadata[nid] = {}
                node_ids.add(nid)

    return nodes, edges, metadata

File: playground-server/graph/scala-parse.js
/** Extract individual tuple strings from Scala `List(...)` output. */
export function extractScalaTuples(stdout) {
  if (!stdout) return [];
  let idx = stdout.indexOf('= List(');
  let contentStart;
  if (idx !== -1) {
    contentStart = idx + 7;
  } else {
    idx = stdout.indexOf('List(');
    if (idx === -1) return [];
    contentStart = idx + 5;
  }
  let depth = 1;
  let pos = contentStart;
  while (pos < stdout.length && depth > 0) {
    if (stdout[pos] === '(') depth += 1;
    else if (stdout[pos] === ')') depth -= 1;
    pos += 1;
  }
  const content = stdout.slice(contentStart, pos - 1);
  const tuples = [];
  let i = 0;
  while (i < content.length) {
    if (content[i] === '(') {
      let d = 1;
      let j = i + 1;
      while (j < content.length && d > 0) {
        if (content[j] === '(') d += 1;
        else if (content[j] === ')') d -= 1;
        j += 1;
      }
      if (d === 0) {
        tuples.push(content.slice(i, j));
        i = j;
        continue;
      }
    }
    i += 1;
  }
  return tuples;
}

/** Split a Scala tuple string by top-level commas into fields. */
export function splitScalaTuple(tupleStr) {
  let inner = tupleStr;
  if (inner.startsWith('(') && inner.endsWith(')')) inner = inner.slice(1, -1);
  const fields = [];
  let current = '';
  let depth = 0;
  let inString = false;
  let escaped = false;
  for (let i = 0; i < inner.length; i++) {
    const ch = inner[i];
    if (escaped) {
      current += ch;
      escaped = false;
      continue;
    }
    if (ch === '\\') {
      current += ch;
      escaped = true;
      continue;
    }
    if (ch === '"') {
      inString = !inString;
      current += ch;
      continue;
    }
    if (inString) {
      current += ch;
      continue;
    }
    if ('([{'.includes(ch)) {
      depth += 1;
      current += ch;
      continue;
    }
    if (')]}'.includes(ch)) {
      depth -= 1;
      current += ch;
      continue;
    }
    if (ch === ',' && depth === 0) {
      fields.push(current.trim());
      current = '';
      continue;
    }
    current += ch;
  }
  if (current) fields.push(current.trim());
  return fields;
}

/** Parse a single Scala value into JS. */
export function parseScalaField(field) {
  let f = field.trim();
  if (!f) return null;
  if (f.endsWith('L') && f.length > 1) {
    const rest = f.slice(0, -1);
    if (/^-?\d+$/.test(rest)) f = rest;
  }
  if (f.startsWith('"') && f.endsWith('"') && f.length >= 2) {
    return f.slice(1, -1).replace(/\\"/g, '"').replace(/\\\\/g, '\\');
  }
  if (f === 'None') return null;
  if (f.startsWith('Some(') && f.endsWith(')')) {
    const innerVal = f.slice(5, -1).trim();
    const eqIdx = innerVal.indexOf(' = ');
    if (eqIdx !== -1) {
      const numStr = innerVal.slice(eqIdx + 3).trim();
      const n = Number(numStr);
      return Number.isNaN(n) ? numStr : n;
    }
    const n = Number(innerVal);
    return Number.isNaN(n) ? innerVal : n;
  }
  const n = Number(f);
  return Number.isNaN(n) ? f : n;
}

/** Parse 6-field metadata tuples into { nodeId: metadata }. */
export function parseMetadataTuples(stdout) {
  const metadata = {};
  for (const t of extractScalaTuples(stdout)) {
    const fields = splitScalaTuple(t);
    if (fields.length < 6) continue;
    try {
      const fid = parseScalaField(fields[0]);
      const code = parseScalaField(fields[1]);
      const lineNum = parseScalaField(fields[2]);
      const colNum = parseScalaField(fields[3]);
      const order = parseScalaField(fields[4]);
      const nodeType = parseScalaField(fields[5]);
      metadata[String(fid)] = {
        code: typeof code === 'string' ? code : code != null ? String(code) : '',
        line_number: lineNum,
        column_number: colNum,
        order: order != null ? order : -1,
        argument_index: -1,
        node_type: typeof nodeType === 'string' ? nodeType : '',
      };
    } catch {
      continue;
    }
  }
  return metadata;
}

/** Parse 7-field AST tuples into (nodes, edges, metadata). */
export function parseAstTuples(stdout) {
  const nodes = [];
  const edges = [];
  const metadata = {};
  for (const t of extractScalaTuples(stdout)) {
    const fields = splitScalaTuple(t);
    if (fields.length < 7) continue;
    try {
      const fid = parseScalaField(fields[0]);
      const code = parseScalaField(fields[1]);
      const lineNum = parseScalaField(fields[2]);
      const colNum = parseScalaField(fields[3]);
      const order = parseScalaField(fields[4]);
      const nodeType = parseScalaField(fields[5]);
      const parentId = parseScalaField(fields[6]);
      const nodeIdStr = String(fid);
      nodes.push({
        id: nodeIdStr,
        label: typeof nodeType === 'string' ? nodeType : nodeType != null ? String(nodeType) : '',
      });
      metadata[nodeIdStr] = {
        code: typeof code === 'string' ? code : code != null ? String(code) : '',
        line_number: lineNum,
        column_number: colNum,
        order: order != null ? order : -1,
        argument_index: -1,
        node_type: typeof nodeType === 'string' ? nodeType : '',
      };
      if (parentId != null) {
        edges.push({ source: String(parentId), target: nodeIdStr, label: '' });
      }
    } catch {
      continue;
    }
  }

  const nodeIds = new Set(nodes.map((n) => n.id));
  for (const e of edges) {
    for (const nid of [e.source, e.target]) {
      if (!nodeIds.has(nid)) {
        nodes.push({ id: nid, label: '', shape: '' });
        metadata[nid] = {};
        nodeIds.add(nid);
      }
    }
  }

  return { nodes, edges, metadata };
}
